### PR TITLE
AmazonPay as a Tradittional Payment Option - Product Page and Mini Cart Placements - Support Alexa Delivery notifications

### PR DESCRIPTION
**Pull request by dpanta94**
### All Submissions:

* [x] Does your code follow the [Extendables](https://extendomattic.wordpress.com/standardizations/) standards?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

#### Introduce Amazon Pay as a traditional WooCommerce Gateway Option.
* Adds Amazon Pay as a tradditional gateway option. [SauCal pull#145](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/145)
* Adds Amazon Pay "classic" support for the WooCommerce Checkout Block. [SauCal pull#155](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/155)
* Makes a phone to be supplied required, when using Amazon Pay "classic". [SauCal pull#157](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/157)
* Fixes a possible overwrite. [SauCal pull#159](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/159)

Introduces Amazon Pay as a traditional WooCommerce Gateway based on an option in the plugin settings, which we will be referring to as 'classic'.

Key differences when using the classic option is we need to provide to Amazon the shipping details of the order if the order needs shipping and the customer doesn't have the chance to review the order after being redirected to Amazon since he is going to complete his checkout process there and then he will return to merchant's site.

In the order-pay endpoint, we had to change where the already implemented checkout button was appearing in order to put the classic method in its place. Its still available, but now appearing in the same place as in the checkout page.

Technical Details

We have implemented the integration named "Additional payment button" by Amazon.
@see https://developer.amazon.com/docs/amazon-pay-apb-checkout/additional-payment-button-overview.html

In the checkout endpoint, when Amazon classic is selected as the payment method. Gateway's process_payment method
returns a result object containing all the required params for the JS to create an Amazon button to initiate the payment, while also providing a redirect value to a HTML element of the same page that does not exists so that JS will have the chance to initiate the Amazon process. JS hooks on the 'checkout_place_order_success", creates the Amazon button hidden and then triggers a click on it for the process to be initiated. That way the flow appears to the end customer as a one-step process.

In the order-pay endpoint, when Amazon classic is selected as the payment method. We overwrite the form's default functionality to an AJAX request in our own endpoint in order to follow the same process as in the checkout endpoint.

Provides the functionality for integrating Amazon Pay "Classic" on the checkout block of WooCommerce blocks.

#### Adds support for Amazon Pay in WooCommerce's mini cart
* Adds support for Amazon Pay placement in the mini-cart. [SauCal pull#146](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/146)
* Fixes button's margins. [SauCal pull#161](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/161)
* Refreshes data on wc_fragments_refreshed too. [SauCal pull#163](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/163)

Adds support for the Amazon Pay button in the WooCommerce's mini cart.

#### Adds support for Amazon Pay in Product Pages
* Adds support for the Amazon Pay button in Product Pages.. [SauCal pull#147](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/147)
* CSS and output amends. [SauCal pull#160](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/160)

Adds support for the Amazon Pay button in Product Pages.

Technical details

When the amazon pay button located in the product page is clicked an AJAX request is taking place which saves the active cart
in WooCommerce's session, empties current cart and adds only the selected product and finally provides the JS with required parameters for amazon to initiate the new checkout session.

When the checkout session is completed, the cart that was stored gets restored.

#### Adds support for Alexa Delivery notifications
* Adds support for Alexa Delivery Notifications. [SauCal pull#148](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/148)

Adds support for Alexa Delivery Notifications. Shipping plugins can simply call

`do_action( 'woocommerce_amazon_pa_enable_alexa_notifications', $tracking_number, $carrier, $order_id );`

in a time where wc_get_order is applicable to be called, and as long as the merchant has enabled the corresponding option, alexa delivery notifications will be enabled for the $order_id.

##### Fixes Gateway's availability determination
* [SauCal pull#151](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/151)
* [SauCal pull#154](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/154)
##### Update package.json
* [SauCal pull#156](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/156)
##### Honor subscription support option for all placements
* [SauCal pull#158](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/158)
##### Fixes Naming convections / Typo / PHPCS
* [SauCal pull#162](https://github.com/saucal/woocommerce-gateway-amazon-pay/pull/162)
### Other information:
This is a PR consisting of 15 smaller PRs (all links provided).
Main purpose is to introduce 4 new features.
- Amazon Pay as a tradittional payment option support
- Amazon Pay in the mini-cart placement support
- Amazon Pay in the product page placement support
- Alexa delivery notifications support

* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changelog entry

> Add - Introduces Amazon Pay as a traditional WooCommerce Gateway.
> Add - 3 new options. For controlling the traditional appearance of Amazon Pay, gateway's title and description.
> Add - Supports Amazon Pay "Classic" on the checkout block of WooCommerce Blocks.
> Add - Support for Amazon Pay in WooCommerce's mini cart.
> Add - Support for Amazon Pay button on Product pages.
> Add - Support for Alexa Delivery Notifications

### CODE REVIEW

File: src/renderAmazonButton.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * @param {string} buttonSettingsFlag Specifies the context of the rendering.
+ * @param {string} checkoutConfig The checkoutConfig with which we will provide Amazon Pay.
+ * @returns {object} The Amazon Pay rendered button.
+ */
+const renderAmazonButton = ( btnId, buttonSettingsFlag, checkoutConfig ) => {
+	let amazonPayBtn = null;
+	const btns = document.querySelectorAll( btnId );
+	for ( const btn of btns ) {
+		const thisId = '#' + btn.getAttribute( 'id' );
+		const button_settings = getButtonSettings(
```
Comment (kalessil) [inline src/renderAmazonButton.js:44]: We have inconsistent naming conventions here.

File: src/utils.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/**
+ * Returns an array of the sibling of the elem that have the class className.
+ * 
+ * @param {node} elem The element's whose siblings we are searching for.
+ * @param {string} className Filter the siblings by this className.
+ * @returns {array}
+ */
+export const getSiblings = ( elem, className ) => {
```
Comment (kalessil) [inline src/utils.js:8]: Also shortenings.

File: src/utils.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+};
+
+/**
+ * Returns the backend provided settings based on the name param.
+ *
+ * @param {string} name The settings to access.
+ * @returns {object}|{null}
+ */
+export const getBlocksConfiguration = ( name ) => {
+	const amzPayServerData = wc.wcSettings.getSetting( name, null );
```
Comment (kalessil) [inline src/utils.js:33]: Also shortenings.

File: assets/js/amazon-wc-checkout.js [hunk 4]
```diff
@@ -51,18 +195,19 @@
...
 		renderButton();
 		$( document.body ).on( 'updated_wc_div', renderButton );
 		$( document.body ).on( 'updated_checkout', renderButton );
 		$( document.body ).on( 'payment_method_selected', renderButton );
 		$( document.body ).on( 'updated_shipping_method', renderButton );
 
-		function attemptRefreshData() {
-			var dataCont = $( '#wc-apa-update-vals' );
+		function attemptRefreshData( flag ) {
+			var dataCont = 'cart' === flag ? $( '#wc-apa-update-vals-cart' ) : $( '#wc-apa-update-vals' );
```
Comment (kalessil) [inline assets/js/amazon-wc-checkout.js:210]: Also shortening?

File: assets/js/amazon-wc-checkout.js [hunk 5]
```diff
@@ -1,40 +1,184 @@
...
-					sandbox: amazon_payments_advanced.sandbox === '1' ? true : false,
-					// customize the buyer experience
-					productType: amazon_payments_advanced.action,
-					placement: amazon_payments_advanced.placement,
-					buttonColor: amazon_payments_advanced.button_color,
-					checkoutLanguage: amazon_payments_advanced.button_language !== '' ? amazon_payments_advanced.button_language.replace( '-', '_' ) : undefined,
-					// configure Create Checkout Session request
-					createCheckoutSessionConfig: amazon_payments_advanced.create_checkout_session_config
-				};
+				var button_settings = getButtonSettings( buttonSettingsFlag );
```
Comment (kalessil) [inline assets/js/amazon-wc-checkout.js:181]: We have inconsistent naming conventions here.

File: assets/js/amazon-wc-checkout.js [hunk 6]
```diff
@@ -1,40 +1,184 @@
...
+/*global amazon_payments_advanced, amazon, wc_checkout_params */
 ( function( $ ) {
 	$( function() {
 		var button_count = 0;
 		var button_id = '#pay_with_amazon';
-		function renderButton() {
-			attemptRefreshData();
-			if ( 0 === $( button_id ).length ) {
+		var classic_button_id = '#classic_pay_with_amazon';
+		var amzCreateCheckoutConfig = null;
```
Comment (kalessil) [inline assets/js/amazon-wc-checkout.js:7]: Also, we have inconsistent naming style here.

**Review by kalessil — CHANGES_REQUESTED** [review]
Did the first review round and I like the overall direction here.

I left my comments and the most of them are around:
- not using shortenings like `btn`, `amzn`, `elem`
- naming conventions consistency (in PHP we use Snake case and in React components Camel case)

Consistency is also important, so in some parts of JS where Snake case is already used, we can keep using it.

**Comment by dpanta94** [comment]
Hey @kalessil ! We have made amends for all your comments above. Thank you!

**Review by kalessil — APPROVED** [review]
LGTM!

**Pull request merged**